src/BaseControllerV2.ts

```diff
@@ -195,17 +156,18 @@ export class BaseController<
    * @param callback - Callback for updating state, passed a draft state
    *   object. Return a new state object or mutate the draft to update state.
    */
-  protected update(
-    callback: (state: Draft<IsJsonable<S>>) => void | IsJsonable<S>,
-  ) {
-    const [nextState, patches] = produceWithPatches(
-      this.internalState,
-      callback,
-    );
-    this.internalState = nextState as IsJsonable<S>;
+  protected update(callback: (state: Draft<S>) => void | S) {
+    // We run into ts2589, "infinite type depth", if we don't cast
+    // produceWithPatches here.
```

Comment:
Previously we had avoided this problem by casting one line down from here. That way we only had to cast `nextState`, rather than the while `produceWithPatches` function. Would that still work here? :thinking:

Reply:
Hm, I'm not sure if we're referring to the same problem. Casting something other than `produceWithPatches` itself doesn't help, because that's what TypeScript is complaining about. See:

![image](https://user-images.githubusercontent.com/25517051/123491095-c29b4d80-d5ca-11eb-8b9d-046e1f89b5ad.png)

Reply:
Oh, I see. You're right, I was thinking of a different issue. This issue you're running into here is an incompatibility between `Patches[]` and `Json`  (or any recursive type possibly?). Beforehand it wasn't an issue because `internalState` was of type `Record<string, unknown>`. This makes sense then.